### PR TITLE
fix and unit-tests for issue #214

### PR DESCRIPTION
**Pull request by vijayvikrant**

### CODE REVIEW

File: netmaster/master/netmaster_test.go [hunk 1]
```diff
@@ -296,3 +308,33 @@ func TestVxlanConfigWithLateHostBindings(t *testing.T) {
...
+            "Name"              : "orange",
+            "PktTag"            : 1300
+        }]
+    }]}`)
+
+	initFakeStateDriver(t)
+	defer deinitFakeStateDriver()
+
+	// Create the Tenant and Network
+	applyConfig(t, cfgBytes)
```
Comment (jainvipin) [inline netmaster/master/netmaster_test.go]: can you also confirm if this also returns the error? 
Just not establishing the state is not sufficient perhaps for a -ve test case

File: netmaster/master/netmaster_test.go [hunk 2]
```diff
@@ -296,3 +308,33 @@ func TestVxlanConfigWithLateHostBindings(t *testing.T) {
...
+	fakeDriver.DumpState()
+
+	keys := []string{"tenant-one"}
+	verifyKeys(t, keys)
+
+	keys = []string{"nets/orange"}
+	fakeDriver.DumpState()
```
Comment (jainvipin) [inline netmaster/master/netmaster_test.go]: will the above print a lot of gunk during test runs? Can we do that only when test case fails, or better yet remove it all together.

File: netmaster/master/netmaster_test.go (hunk 1 above)
Comment (vijayvikrant) [inline netmaster/master/netmaster_test.go]: As of today create Network will not return an error for this. So, what you are saying is that I check for err nonetheless, so that when we fix it tomorrow, that also gets verified?

File: netmaster/master/netmaster_test.go (hunk 2 above)
Comment (vijayvikrant) [inline netmaster/master/netmaster_test.go]: will remove DumpState statements.

File: netmaster/master/netmaster_test.go (hunk 1 above)
Comment (jainvipin) [inline netmaster/master/netmaster_test.go]: @vijayvikrant - hmm.. if it doesn't return the error, how do we make sure that in reality it returns error as well (say from netctl/contivctl)?
> As of today create Network will not return an error for this. So, what you are saying is that I check for err nonetheless, so that when we fix it tomorrow, that also gets verified?

Comment (vijayvikrant) [inline netmaster/master/netmaster_test.go]: @jainvipin - I meant as of today CreateNetwork does not return an error. I am making changes to CreateNetwork to return error when the specified vxlan/vlan is out of allowed range for the tenant.. And also make the test case check for the error.

Comment (jainvipin) [inline netmaster/master/netmaster_test.go]: @vijayvikrant - got it... it seems like you plan to modify this test case once you fix issue; up to you if you want to combine this with other PR or keep them separate.
otherwise LGTM

**Comment by vijayvikrant** [comment]
@jainvipin so, this issue was also present in our sanity script. I have modified the sanity code to make sure that the vlans/vxlans we allocate are in the tenant's configured ranges. The sanity passes after this change.

File: utils/netutils/netutils.go [hunk 3]
```diff
@@ -213,6 +213,23 @@ func ParseTagRanges(ranges string, tagType string) ([]TagRange, error) {
 	return tagRanges, nil
 }
 
+// IsTagInRange verifies if the passed pktTag adheres to the valueRange
+func IsTagInRange(pktTag int, valueRangeStr string, encap string) bool {
+	// Parse the valueRange to form a []TagRange object
+	valueRanges, err := ParseTagRanges(valueRangeStr, "vlan")
+	if err != nil {
+		log.Errorf("Not able to pass valueRange. Err: %s", err)
```
Comment (jainvipin) [inline utils/netutils/netutils.go]: minor typo: `parse` instead of `pass`

File: utils/netutils/netutils.go [hunk 4]
```diff
@@ -213,6 +213,23 @@ func ParseTagRanges(ranges string, tagType string) ([]TagRange, error) {
...
+		return false
+	}
+
+	for _, valueRange := range valueRanges {
```
Comment (jainvipin) [inline utils/netutils/netutils.go]: would you think it is better to this change in the ParseTagRanges itself, then we can be sure anyplace we parse the tagrange this is validated.

Comment (vijayvikrant) [inline utils/netutils/netutils.go]: @jainvipin not sure I get it. The ranges are created in CreateTenant and so it makes sense to call  ParseTagRange there. CreateNetwork needs to check if the provided vlan/vxlan fall into the configured range. As per your comment CreateNetwork should do ParseTagRanges(Range, encap, vlan/vxlan) - which should validate if PktTag is in range and return multi-values (range, bool)?

Comment (vijayvikrant) [inline utils/netutils/netutils.go]: @jainvipin a lot of places where we would parse tagRange, we won't have the Tag information yet. Like in the earlier example, CreateTenant will parse tagRange, but not have pktTag info yet. That would come later when netctl net create happens.. Am I missing something?

Comment (jainvipin) [inline utils/netutils/netutils.go]: @vijayvikrant - as we talked offline, let's see if the call to ParseRange can ba avoided in IsTagInRange() i.e. pass the range from pre-parsed (tenant) to avoid doing this twice

File: netmaster/master/network.go [hunk 5]
```diff
@@ -236,11 +276,18 @@ func CreateNetwork(network intent.ConfigNetwork, stateDriver core.StateDriver, t
 		nwCfg.ExtPktTag = int(extPktTag)
 		nwCfg.PktTag = int(pktTag)
 	} else if network.PktTagType == "vxlan" {
+		if !IsTagInRange(network.PktTag, "vxlan", gCfg.Auto.VXLANs, stateDriver, tenantName) {
+			return fmt.Errorf("vxlan %d does not adhere to tenant's vxlan range %s", network.PktTag, gCfg.Auto.VXLANs)
+		}
+
 		nwCfg.ExtPktTag = network.PktTag
 		nwCfg.PktTag = network.PktTag
```
Comment (jainvipin) [inline netmaster/master/network.go]: @vvb - this line is still a problem.
The changes you made are needed, however not sufficient if we want to allow vxlan range beyond 4k. for example:
```
                } else if nwCfg.PktTagType == "vxlan" {
                        extPktTag, pktTag, err = gCfg.AllocVXLAN(rm)
                        if err != nil {
                                return err
                        }
                }
```
The following code in auto-allocation allows for both vlan-tag and vxlan-tag to be assigned; we need to follow the similar model for the code above to avoid a conflict and accessing beyond the range.

Bottom line: if you try a vxlan id beyond 4k range, it might still not work even with all your changes. In order to fix this you'd have to allocate a vlan-tag for it that is still in the allowed range. If you follow ofnet.AddVlan() call from ovsdriver.go for CreateNetwork() may be it will all make sense.

I understand you had not added all this, but let's fix this all of them.

**Comment by vijayvikrant** [comment]
@jainvipin Yes, the plan is to do those changes as a part of PR for #216. This one is for #214 only.

**Comment by jainvipin** [comment]
Thanks @vijayvikrant 

LGTM

**Comment by jainvipin** [comment]
ping @mapuri 
can you also look at the diffs and confirm the changes?

File: netmaster/master/network.go [hunk 6]
```diff
@@ -175,6 +176,45 @@ func deleteDockNet(tenantName, networkName, serviceName string) error {
 	return nil
 }
 
+// IsTagInRange verifies if the requested vlan/vxlan adheres to tenants configured range
+func IsTagInRange(pktTag int, pktTagType string, pktTagRange string, stateDriver core.StateDriver, tenantName string) bool {
```
Comment (mapuri) [inline netmaster/master/network.go]: does this function need to be exported/public? Otherwise we can start it with lower-case.

Also can pktTag be negative? I am wondering if the `pktTag` argument to this function can be a uint instead to avoid the uint type cast below, otherwise we need to deal with testing input for negative numbers as well.

File: netmaster/master/netmaster_test.go [hunk 7]
```diff
@@ -296,3 +308,164 @@ func TestVxlanConfigWithLateHostBindings(t *testing.T) {
...
+
+	tenant := cfg.Tenants[0]
+	err = CreateTenant(fakeDriver, &tenant)
+	if err != nil {
+		t.Fatalf("error '%s' creating tenant\n", err)
+	}
+
+	err = CreateNetworks(fakeDriver, &tenant)
+	if shouldFail {
+		if err == nil {
```
Comment (mapuri) [inline netmaster/master/netmaster_test.go]: it might be good to test that err returned is indeed for `OutOfRange` as `CreateNetworks()` might return other errors as well?

Comment (vijayvikrant) [inline netmaster/master/netmaster_test.go]: @mapuri I have made the changes for this. PTAL

File: netmaster/master/netmaster_test.go [hunk 8]
```diff
@@ -296,3 +309,173 @@ func TestVxlanConfigWithLateHostBindings(t *testing.T) {
...
+
+		var expError string
+		network := tenant.Networks[0]
+		if network.PktTagType == "vlan" {
+			expError = fmt.Sprintf("vlan %d does not adhere to tenant's vlan range %s", network.PktTag, tenant.VLANs)
+		} else {
+			expError = fmt.Sprintf("vxlan %d does not adhere to tenant's vxlan range %s", network.PktTag, tenant.VXLANs)
+		}
+
+		if err.Error() != expError {
```
Comment (mapuri) [inline netmaster/master/netmaster_test.go]: this might panic if err is nil, so may be we should check for `err == nil` as well?

**Comment by mapuri** [comment]
thanks @vijayvikrant , LGTM other than one minor comment.

File: netmaster/master/netmaster_test.go (hunk 8 above)
Comment (vijayvikrant) [inline netmaster/master/netmaster_test.go]: @mapuri that is correct. I will change it.

**Comment by vijayvikrant** [comment]
@mapuri added the check.
